common/hobbitlaunch.8
.TH HOBBITLAUNCH 8 "Version 4.2.3:  4 Feb 2009" "Xymon"
.SH NAME
hobbitlaunch \- Master program to launch other Xymon programs

.SH SYNOPSIS
.B "hobbitlaunch [options]"

.SH DESCRIPTION
.I hobbitlaunch(8)
is the main program that controls the execution and scheduling of 
all of the components in the Xymon system.

hobbitlaunch allows the administrator to add, remove or change the set of
Xymon applications and extensions without restarting Xymon -
hobbitlaunch will automatically notice any changes in the set of tasks,
and change the scheduling of activities accordingly.

hobbitlaunch also allows the administrator to setup specific logfiles for
each component of the Xymon system, instead of getting output
from all components logged to a single file.

.SH OPTIONS
.IP "--env=FILENAME"
Loads the environment from FILENAME before starting other tools.
The environment defined by FILENAME is the default, it can be
overridden by the ENVFILE option in 
.I hobbitlaunch.cfg(5)

.IP "--config=FILENAME"
This option defines the file that hobbitlaunch scans for tasks it
must launch. A description of this file is in 
.I hobbitlaunch.cfg(5)
The default tasklist is /etc/hobbitlaunch.cfg

.IP "--log=FILENAME"
Defines the logfile where hobbitlaunch logs information about
failures to launch tasks and other data about the operation of
hobbitlaunch. Logs from individual tasks are defined in the hobbitlaunch.cfg
file. By default this is logged to stdout.

.IP "--pidfile=FILENAME"
Filename which hobbitlaunch saves its own process-ID to. Commonly used by
automated start/stop scripts.

.IP "--verbose"
Logs the launch of all tasks to the logfile. Note that the logfile
may become quite large if you enable this.

.IP "--dump"
Just dump the contents of the hobbitlaunch.cfg file after parsing it.
Used for debugging.

.IP "--debug"
Enable debugging output while running.

.IP "--no-daemon"
hobbitlaunch normally detaches from the controlling tty and runs as a
background task. This option keeps it running in the foreground.

.SH STARTING TASKS
hobbitlaunch will read the configuration file and start all of the tasks
listed there.

If a task completes abnormally (i.e. terminated by a signal or with
a non-zero exit status), then hobbitlaunch will attempt to restart it
5 times. If it still will not run, then the task is disabled for 10
minutes. This will be logged to the hobbitlaunch logfile.

If the configuration file changes, hobbitlaunch will re-read it and 
notice any changes. If a running task was removed from the configuration,
then the task is stopped. If a new task was added, it will be started.
If the command used for a task changed, or it was given a new environment
definition file, or the logfile was changed, then the task is stopped
and restarted with the new definition.

.SH "SEE ALSO"
hobbitlaunch.cfg(5), xymon(7)



Changes
Changes from 4.2.3 -> 4.3.0 (09 Feb 2008)
=========================================
Core changes:
* New API's for loadhosts and sendmessage, in preparation for
  the full 5.0 changes.
* Always use getcurrenttime() instead of time().
* Support for defining holidays as non-working days in alerts and
  SLA calculations.
* Hosts which appear on multiple pages in the web display can
  use any page they are on in the alerting rules and elsewhere.
* Worker modules (RRD, client-data parsers etc) can operate on
  remote hosts from the hobbitd daemon, for load-sharing.
* Various bugfixes collected over time.

Network test changes:
* Merged new network tests from trunk: SOAP-over-HTTP,
  SSL minimum cipher strength
* Changed network test code to always report a validity period
  for network tests, so it it possible to run network tests less
  often than every 30 minutes (e.g. once an hour).
* Make the content-type setting in HTTP POST tests configurable.
* Make the source-address used for TCP tests configurable.
* Make the acceptable HTTP result codes configurable.
* Use and save HTTP session cookies.

Web changes
* Support generic drop-down lists in templates.
* "NOCOLUMNS" changed to work for all columns.
* New "group-sorted" definition to auto-sort hosts in a group
* Use browser tooltips for host comments
* "Compact" status allows several statuses to appear as a single
  status on the overview webpages.
* Trends page can select the time period to show. Buttons provided
  for the common selections.
* Ghost list report now lists possible candidates for a ghost, 
  based on IP-address or unqualified hostname.

Report changes
* Number of outages as SLA parameter

Miscellaneous
* hobbitlaunch support for running tasks only on certain hosts,
  and for a maximum time.
* Alert script get a unique ID for each alert.



Changes from 4.2.2 -> 4.2.3 (09 Feb 2008)
=========================================
* Time-out code changed to use clock_gettime() with CLOCK_MONOTONIC
* Bugfix for hobbitd/hobbitd_worker communication going out-of-sync
  resulting in "garbled data" being logged and worker modules 
  stopping.
* NCV module now works with negative numbers.
* Several bugfixes in DNS lookup code - could lead to crashes when
  performing DNS tests.
* Switch to C-ARES 1.6.0 - drop support for older versions.
* Run more TCP tests in parallel by not waiting for very slow
  connections to complete before starting new ones.
* Added "hostlist" web utility for spreadsheet-reporting of the
  hosts in Hobbit.


Changes from 4.2.0 -> 4.2.2 (01 Dec 2008)
=========================================

Changelog extracted from Subversion.


------------------------------------------------------------------------
r5935 | storner | 2008-11-26 12:57:11 +0100 (Wed, 26 Nov 2008) | 3 lines
Changed paths:
   A /branches/4.2.0 (from /trunk:5040)

Create branch for maintenance of the 4.2.x release.


------------------------------------------------------------------------
r5936 | storner | 2008-11-26 13:13:50 +0100 (Wed, 26 Nov 2008) | 3 lines
Changed paths:
   M /branches/4.2.0/common/bb-hosts.5

The ability to set a DOWNTIME for an individual test was mentioned in the Changes file, but not documented in the bb-hosts man-page.


------------------------------------------------------------------------
r5937 | storner | 2008-11-26 13:14:48 +0100 (Wed, 26 Nov 2008) | 3 lines
Changed paths:
   M /branches/4.2.0/hobbitd/hobbitd_client.c

The hobbitd_client program would crash when running in test-mode, i.e. launched with the --test option to test disk-, procs- and ports-settings. This only applies to running it in test-mode, normal operation is not affected.


------------------------------------------------------------------------
r5938 | storner | 2008-11-26 13:15:40 +0100 (Wed, 26 Nov 2008) | 3 lines
Changed paths:
   M /branches/4.2.0/lib/headfoot.c
   M /branches/4.2.0/web/bb-datepage.c
   M /branches/4.2.0/web/bb-eventlog.c
   M /branches/4.2.0/web/bb-findhost.c
   M /branches/4.2.0/web/bb-rep.c
   M /branches/4.2.0/web/bb-snapshot.c
   M /branches/4.2.0/web/hobbit-enadis.c
   M /branches/4.2.0/web/hobbit-nkedit.c

Some of the web page would generate an extra line with "Content-type: text/html" at the top of the webpage, although this would most often be hidden by the menubar.


------------------------------------------------------------------------
r5939 | storner | 2008-11-26 13:16:21 +0100 (Wed, 26 Nov 2008) | 3 lines
Changed paths:
   M /branches/4.2.0/client/runclient.sh

Using extra options with the client-side "runclient.sh" script when performing a restart command fails to pass the extra options to the final command. 


------------------------------------------------------------------------
r5940 | storner | 2008-11-26 13:17:12 +0100 (Wed, 26 Nov 2008) | 3 lines
Changed paths:
   M /branches/4.2.0/hobbitd/hobbitfetch.c

hobbitfetch insisted that hosts must have a valid IP-adress in bb-hosts. This causes problems for hosts with dynamic IP-adresses. This patch causes hobbitfetch to lookup the IP-address at run-time if the IP is listed as "0.0.0.0".


------------------------------------------------------------------------
r5941 | storner | 2008-11-26 13:20:44 +0100 (Wed, 26 Nov 2008) | 3 lines
Changed paths:
   M /branches/4.2.0/web/hobbit-confreport.c
   M /branches/4.2.0/web/hobbitsvc-info.c


common/hobbitlaunch.c
/*----------------------------------------------------------------------------*/
/* Hobbit application launcher.                                               */
/*                                                                            */
/* This is used to launch various parts of the Hobbit system. Some programs   */
/* start up once and keep running, other must run at various intervals.       */
/*                                                                            */
/* Copyright (C) 2004-2009 Henrik Storner <henrik@hswn.dk>                    */
/*                                                                            */
/* This program is released under the GNU General Public License (GPL),       */
/* version 2. See the file "COPYING" for details.                             */
/*                                                                            */
/*----------------------------------------------------------------------------*/

static char rcsid[] = "$Id: hobbitlaunch.c,v 1.41 2006-07-20 16:06:41 henrik Exp $";

#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <sys/time.h>
#include <sys/wait.h>
#include <time.h>
#include <unistd.h>
#include <errno.h>
#include <ctype.h>
#include <regex.h>

#include "libbbgen.h"

#include <signal.h>

/*
 * config file format:
 *
 * [hobbitd]
 * 	CMD hobbitd --no-daemon
 * 	LOGFILE /var/log/hobbitd.log
 *
 * [bbdisplay]
 * 	CMD bb-display.sh
 * 	INTERVAL 5m
 */

#define MAX_FAILS 5

typedef struct grouplist_t {
	char *groupname;
	int currentuse, maxuse;
	struct grouplist_t *next;
} grouplist_t;

typedef struct tasklist_t {
	char *key;
	int disabled;
	grouplist_t *group;
	char *cmd;
	int interval, maxruntime;
	char *logfile;
	char *envfile, *envarea, *onhostptn;
	pid_t pid;
	time_t laststart;
	int exitcode;
	int failcount;
	int cfload;	/* Used while reloading a configuration */
	int beingkilled;
	struct tasklist_t *depends;
	struct tasklist_t *next;
} tasklist_t;
tasklist_t *taskhead = NULL;
tasklist_t *tasktail = NULL;
grouplist_t *grouphead = NULL;

volatile time_t nextcfgload = 0;
volatile int running = 1;
volatile int dologswitch = 0;

void update_task(tasklist_t *newtask)
{
	tasklist_t *twalk;
	int freeit = 1;
	int logfilechanged = 0;
	int envfilechanged = 0;
	int disablechanged = 0;

	for (twalk = taskhead; (twalk && (strcmp(twalk->key, newtask->key))); twalk = twalk->next);

	if (twalk) {
		if (twalk->logfile && newtask->logfile) {
			logfilechanged = strcmp(twalk->logfile, newtask->logfile);
		}
		else if (twalk->logfile || newtask->logfile) {
			logfilechanged = 1;
		}
		else {
			logfilechanged = 0;
		}

		if (twalk->envfile && newtask->envfile) {
			envfilechanged = strcmp(twalk->envfile, newtask->envfile);
		}
		else if (twalk->envfile || newtask->envfile) {
			envfilechanged = 1;
		}
		else {
			envfilechanged = 0;
		}

		if (twalk->envarea && newtask->envarea) {
			envfilechanged = strcmp(twalk->envarea, newtask->envarea);
		}
		else if (twalk->envarea || newtask->envarea) {
			envfilechanged = 1;
		}
		else {
			envfilechanged = 0;
		}

		if (twalk->disabled != newtask->disabled)
			disablechanged = 1;
		else
			disablechanged = 0;
	}

	if (newtask->cmd == NULL) {
		errprintf("Configuration error, no command for task %s\n", newtask->key);
	}
	else if (twalk == NULL) {
		/* New task, just add it to the list */
		newtask->cfload = 0;

		if (taskhead == NULL) taskhead = newtask;
		else tasktail->next = newtask;

		tasktail = newtask;
		freeit = 0;
	}
	else if (strcmp(twalk->cmd, newtask->cmd) || logfilechanged || envfilechanged || disablechanged) {
		/* Task changed. */
		xfree(twalk->cmd); 
		if (twalk->logfile) xfree(twalk->logfile);
		if (twalk->envfile) xfree(twalk->envfile);
		if (twalk->envarea) xfree(twalk->envarea);
		if (twalk->onhostptn) xfree(twalk->onhostptn);
		twalk->cmd = strdup(newtask->cmd);
		if (newtask->logfile) twalk->logfile = strdup(newtask->logfile); else twalk->logfile = NULL;
		if (newtask->envfile) twalk->envfile = strdup(newtask->envfile); else twalk->envfile = NULL;
		if (newtask->envarea) twalk->envarea = strdup(newtask->envarea); else twalk->envarea = NULL;
		if (newtask->onhostptn) twalk->onhostptn = strdup(newtask->onhostptn); else twalk->onhostptn = NULL;

		/* Must bounce the task */
		twalk->cfload = 1;
	}
	else if (twalk->interval != newtask->interval) {
		twalk->interval = newtask->interval;
		twalk->cfload = 0;
	}
	else if (twalk->maxruntime != newtask->maxruntime) {
		twalk->maxruntime = newtask->maxruntime;
		twalk->cfload = 0;
	}
	else if (twalk->disabled != newtask->disabled) {
		twalk->disabled = newtask->disabled;
		twalk->cfload = 1;
	}
	else {
		/* Task was unchanged */
		twalk->cfload = 0;
	}

	if (freeit) {
		xfree(newtask->key);
		if (newtask->cmd) xfree(newtask->cmd);
		if (newtask->logfile) xfree(newtask->logfile);
		if (newtask->envfile) xfree(newtask->envfile);
		if (newtask->envarea) xfree(newtask->envarea);
		if (newtask->onhostptn) xfree(newtask->onhostptn);
		xfree(newtask);
	}
}

void load_config(char *conffn)
{
	static void *configfiles = NULL;
	tasklist_t *twalk, *curtask = NULL;
	FILE *fd;
	strbuffer_t *inbuf;
	char *p;
	char myhostname[256];

	/* First check if there were no modifications at all */
	if (configfiles) {
		if (!stackfmodified(configfiles)){
			dbgprintf("No files modified, skipping reload of %s\n", conffn);
			return;
		}
		else {
			stackfclist(&configfiles);
			configfiles = NULL;
		}
	}

	errprintf("Loading tasklist configuration from %s\n", conffn);
	if (gethostname(myhostname, sizeof(myhostname)) != 0) {
		errprintf("Cannot get the local hostname, using 'localhost' (error: %s)\n", strerror(errno));
		strcpy(myhostname, "localhost");
	}

	/* The cfload flag: -1=delete task, 0=old task unchanged, 1=new/changed task */
	for (twalk = taskhead; (twalk); twalk = twalk->next) {
		twalk->cfload = -1;
		twalk->group = NULL;
	}

	fd = stackfopen(conffn, "r", &configfiles);
	if (fd == NULL) {
		errprintf("Cannot open configuration file %s: %s\n", conffn, strerror(errno));
		return;
	}

	inbuf = newstrbuffer(0);
	while (stackfgets(inbuf, NULL)) {
		sanitize_input(inbuf, 1, 0); if (STRBUFLEN(inbuf) == 0) continue;

		p = STRBUF(inbuf);
		if (*p == '[') {
			/* New task */
			char *endp;

			if (curtask) {
				update_task(curtask);
				curtask = NULL;
			}

			p++; endp = strchr(p, ']');
			if (endp == NULL) continue;
			*endp = '\0';

			curtask = (tasklist_t *)calloc(1, sizeof(tasklist_t));
			curtask->key = strdup(p);
		}
		else if (curtask && (strncasecmp(p, "CMD ", 4) == 0)) {
			p += 3;
			p += strspn(p, " \t");
			curtask->cmd = strdup(p);
		}
		else if (strncasecmp(p, "GROUP ", 6) == 0) {
			/* Note: GROUP can be used by itself to define a group, or inside a task definition */
			char *groupname;
			int maxuse;
			grouplist_t *gwalk;

			p += 6;
			p += strspn(p, " \t");
			groupname = p;
			p += strcspn(p, " \t");
			if (isdigit((int) *p)) maxuse = atoi(p); else maxuse = 1;

			/* Find or create the grouplist entry */
			for (gwalk = grouphead; (gwalk && (strcmp(gwalk->groupname, groupname))); gwalk = gwalk->next);
			if (gwalk == NULL) {
				gwalk = (grouplist_t *)malloc(sizeof(grouplist_t));
				gwalk->groupname = strdup(groupname);
				gwalk->maxuse = maxuse;
				gwalk->currentuse = 0;
				gwalk->next = grouphead;
				grouphead = gwalk;
			}

			if (curtask) curtask->group = gwalk;
		}
		else if (curtask && (strncasecmp(p, "INTERVAL ", 9) == 0)) {
			char *tspec;
			p += 9;
			curtask->interval = atoi(p);
			tspec = p + strspn(p, "0123456789");
			switch (*tspec) {
			  case 'm': curtask->interval *= 60; break;	/* Minutes */
			  case 'h': curtask->interval *= 3600; break;	/* Hours */
			  case 'd': curtask->interval *= 86400; break;	/* Days */
			}
		}
		else if (curtask && (strncasecmp(p, "MAXTIME ", 8) == 0)) {
			char *tspec;
			p += 8;
			curtask->maxruntime = atoi(p);
			tspec = p + strspn(p, "0123456789");
			switch (*tspec) {
			  case 'm': curtask->maxruntime *= 60; break;	/* Minutes */
			  case 'h': curtask->maxruntime *= 3600; break;	/* Hours */
			  case 'd': curtask->maxruntime *= 86400; break;	/* Days */
			}
		}
		else if (curtask && (strncasecmp(p, "LOGFILE ", 8) == 0)) {
			p += 7;
			p += strspn(p, " \t");
			curtask->logfile = strdup(p);
		}
		else if (curtask && (strncasecmp(p, "NEEDS ", 6) == 0)) {
			p += 6;
			p += strspn(p, " \t");
			for (twalk = taskhead; (twalk && strcmp(twalk->key, p)); twalk = twalk->next);
			if (twalk) {
				curtask->depends = twalk;
			}
			else {
				errprintf("Configuration error, unknown dependency %s->%s", curtask->key, p);
			}
		}
		else if (curtask && (strncasecmp(p, "ENVFILE ", 8) == 0)) {
			p += 7;
			p += strspn(p, " \t");
			curtask->envfile = strdup(p);
		}
		else if (curtask && (strncasecmp(p, "ENVAREA ", 8) == 0)) {
			p += 7;
			p += strspn(p, " \t");
			curtask->envarea = strdup(p);
		}
		else if (curtask && (strcasecmp(p, "DISABLED") == 0)) {
			curtask->disabled = 1;
		}
		else if (curtask && (strncasecmp(p, "ONHOST ", 7) == 0)) {
			regex_t cpattern;
			int status;

			p += 7;
			p += strspn(p, " \t");

			curtask->onhostptn = strdup(p);

			/* Match the hostname against the pattern; if it doesnt match then disable the task */
			status = regcomp(&cpattern, curtask->onhostptn, REG_EXTENDED|REG_ICASE|REG_NOSUB);
			if (status == 0) {
				status = regexec(&cpattern, myhostname, 0, NULL, 0);
				if (status == REG_NOMATCH) curtask->disabled = 1;
			}
			else {
				errprintf("ONHOST pattern '%s' is invalid\n", p);
			}
		}
	}
	if (curtask) update_task(curtask);
	stackfclose(fd);
	freestrbuffer(inbuf);

	/* Running tasks that have been deleted or changed are killed off now. */
	for (twalk = taskhead; (twalk); twalk = twalk->next) {
		switch (twalk->cfload) {
		  case -1:
			/* Kill the task, if active */
			if (twalk->pid) {
				dbgprintf("Killing task %s PID %d\n", twalk->key, (int)twalk->pid);
				twalk->beingkilled = 1;
				kill(twalk->pid, SIGTERM);
			}
			/* And prepare to free this tasklist entry */
			xfree(twalk->key); 
			xfree(twalk->cmd); 
			if (twalk->logfile) xfree(twalk->logfile);
			if (twalk->envfile) xfree(twalk->envfile);
			if (twalk->envarea) xfree(twalk->envarea);
			if (twalk->onhostptn) xfree(twalk->onhostptn);
			break;

		  case 0:
			/* Do nothing */
			break;

		  case 1:
			/* Bounce the task, if it is active */
			if (twalk->pid) {
				dbgprintf("Killing task %s PID %d\n", twalk->key, (int)twalk->pid);
				twalk->beingkilled = 1;
				kill(twalk->pid, SIGTERM);
			}
			break;
		}
	}

	/* First clean out dead tasks at the start of the list */
	while (taskhead->cfload == -1) {
		tasklist_t *tmp;

		tmp = taskhead;
		taskhead = taskhead->next;
		xfree(tmp);
	}

	/* Then unlink and free those inside the list */
	twalk = taskhead;
	while (twalk && twalk->next) {
		tasklist_t *tmp;

		if (twalk->next->cfload == -1) {
			tmp = twalk->next;
			twalk->next = tmp->next;
			xfree(tmp);
		}
		else twalk = twalk->next;
	}

	if (taskhead == NULL) 
		tasktail = NULL;
	else {
		tasktail = taskhead;
		while (tasktail->next) tasktail = tasktail->next;
	}

	/* Make sure group usage counts are correct (groups can change) */
	for (twalk = taskhead; (twalk); twalk = twalk->next) {
		if (twalk->group) twalk->group->currentuse = 0;
	}
	for (twalk = taskhead; (twalk); twalk = twalk->next) {
		if (twalk->group && twalk->pid) twalk->group->currentuse++;
	}
}

void sig_handler(int signum)
{
	switch (signum) {
	  case SIGCHLD:
		break;

	  case SIGHUP:
		nextcfgload = 0;
		dologswitch = 1;
		break;

	  case SIGTERM:
		running = 0;
		break;
	}
}

int main(int argc, char *argv[])
{
	tasklist_t *twalk, *dwalk;
	grouplist_t *gwalk;
	int argi;
	int daemonize = 1;
	int verbose = 0;
	char *config = "/etc/hobbitlaunch.cfg";
	char *logfn = NULL;
	char *pidfn = NULL;
	pid_t cpid;
	int status;
	struct sigaction sa;
	char *envarea = NULL;

	for (argi=1; (argi < argc); argi++) {
		if (strcmp(argv[argi], "--debug") == 0) {
			debug = 1;
		}
		else if (strcmp(argv[argi], "--no-daemon") == 0) {
			daemonize = 0;
		}
		else if (strcmp(argv[argi], "--verbose") == 0) {
			verbose = 1;
		}
		else if (argnmatch(argv[argi], "--config=")) {
			char *p = strchr(argv[argi], '=');
			config = strdup(expand_env(p+1));
		}
		else if (argnmatch(argv[argi], "--log=")) {
			char *p = strchr(argv[argi], '=');
			logfn = strdup(expand_env(p+1));
		}
		else if (argnmatch(argv[argi], "--area=")) {
			char *p = strchr(argv[argi], '=');
			envarea = strdup(p+1);
		}
		else if (argnmatch(argv[argi], "--env=")) {
			char *p = strchr(argv[argi], '=');
			loadenv(p+1, envarea);
		}
		else if (argnmatch(argv[argi], "--pidfile=")) {
			char *p = strchr(argv[argi], '=');
			pidfn = strdup(expand_env(p+1));
		}
		else if (strcmp(argv[argi], "--dump") == 0) {
			/* Dump configuration */

			load_config(config);
			for (gwalk = grouphead; (gwalk); gwalk = gwalk->next) {
				if (gwalk->maxuse > 1) printf("GROUP %s %d\n", gwalk->groupname, gwalk->maxuse);
			}
			printf("\n");
			for (twalk = taskhead; (twalk); twalk = twalk->next) {
				printf("[%s]\n", twalk->key);
				printf("\tCMD %s\n", twalk->cmd);
				if (twalk->disabled) printf("\tDISABLED\n");
				if (twalk->group)    printf("\tGROUP %s\n", twalk->group->groupname);
				if (twalk->depends)  printf("\tNEEDS %s\n", twalk->depends->key);
				if (twalk->interval) printf("\tINTERVAL %d\n", twalk->interval);
				if (twalk->maxruntime) printf("\tMAXTIME %d\n", twalk->maxruntime);
				if (twalk->logfile)  printf("\tLOGFILE %s\n", twalk->logfile);
				if (twalk->envfile)  printf("\tENVFILE %s\n", twalk->envfile);
				if (twalk->envarea)  printf("\tENVAREA %s\n", twalk->envarea);
				if (twalk->onhostptn) printf("\tONHOST %s\n", twalk->onhostptn);
				printf("\n");
			}
			return 0;
		}
	}

	/* Go daemon */
	if (daemonize) {
		pid_t childpid;

		/* Become a daemon */
		childpid = fork();
		if (childpid < 0) {
			/* Fork failed */
			errprintf("Could not fork child\n");
			exit(1);
		}
		else if (childpid > 0) {
			/* Parent exits */
			if (pidfn) {
				FILE *pidfd = fopen(pidfn, "w");

				if (pidfd) {
					fprintf(pidfd, "%d\n", (int)childpid);
					fclose(pidfd);
				}
			}

			exit(0);
		}
		/* Child (daemon) continues here */
		setsid();
	}

	/* If using a logfile, switch stdout and stderr to go there */
	if (logfn) {
		/* Should we close stdin here ? No ... */
		freopen("/dev/null", "r", stdin);
		freopen(logfn, "a", stdout);
		freopen(logfn, "a", stderr);
	}

	save_errbuf = 0;
	setup_signalhandler("hobbitlaunch");
	memset(&sa, 0, sizeof(sa));
	sa.sa_handler = sig_handler;
	sigaction(SIGHUP, &sa, NULL);
	sigaction(SIGTERM, &sa, NULL);
	sigaction(SIGCHLD, &sa, NULL);

	errprintf("hobbitlaunch starting\n");
	while (running) {
		time_t now = getcurrenttime(NULL);

		if (now >= nextcfgload) {
			load_config(config);
			nextcfgload = (now + 30);
		}

		if (logfn && dologswitch) {
			freopen(logfn, "a", stdout);
			freopen(logfn, "a", stderr);
			dologswitch = 0;
		}

		/* Pick up children that have terminated */
		while ((cpid = wait3(&status, WNOHANG, NULL)) > 0) {
			for (twalk = taskhead; (twalk && (twalk->pid != cpid)); twalk = twalk->next);
			if (twalk) {
				twalk->pid = 0;
				twalk->beingkilled = 0;
				if (WIFEXITED(status)) {
					twalk->exitcode = WEXITSTATUS(status);
					if (twalk->exitcode) {
						errprintf("Task %s terminated, status %d\n", twalk->key, twalk->exitcode);
						twalk->failcount++;
					}
					else {
						twalk->failcount = 0;
					}
				}
				else if (WIFSIGNALED(status)) {
					twalk->exitcode = -WTERMSIG(status);
					twalk->failcount++;
					errprintf("Task %s terminated by signal %d\n", twalk->key, abs(twalk->exitcode));
				}

				if (twalk->group) twalk->group->currentuse--;

				/* Tasks that depend on this task should be killed ... */
				for (dwalk = taskhead; (dwalk); dwalk = dwalk->next) {
					if ((dwalk->depends == twalk) && (dwalk->pid > 0)) {
						dwalk->beingkilled = 1;
						kill(dwalk->pid, SIGTERM);
					}
				}
			}
		}

		/* See what new tasks need to get going */
		dbgprintf("\n");
		dbgprintf("Starting tasklist scan\n");
		for (twalk = taskhead; (twalk); twalk = twalk->next) {
			if ((twalk->pid == 0) && !twalk->disabled && (now >= (twalk->laststart + twalk->interval))) {

				if (twalk->depends && ((twalk->depends->pid == 0) || (twalk->depends->laststart > (now - 5)))) {
					dbgprintf("Postponing start of %s due to %s not yet running\n",
						twalk->key, twalk->depends->key);
					continue;
				}

				if (twalk->group && (twalk->group->currentuse >= twalk->group->maxuse)) {
					dbgprintf("Postponing start of %s due to group %s being busy\n",
						twalk->key, twalk->group->groupname);
					continue;
				}

				if ((twalk->failcount > MAX_FAILS) && ((twalk->laststart + 600) < now)) {
					dbgprintf("Releasing %s from failure hold\n", twalk->key);
					twalk->failcount = 0;
				}

				if (twalk->failcount > MAX_FAILS) {
					dbgprintf("Postponing start of %s due to multiple failures\n", twalk->key);
					continue;
				}

				if (twalk->laststart > (now - 5)) {
					dbgprintf("Postponing start of %s, will not try more than once in 5 seconds\n", twalk->key);
					continue;
				}

				dbgprintf("About to start task %s\n", twalk->key);

				twalk->laststart = now;
				twalk->pid = fork();
				if (twalk->pid == 0) {
					/* Exec the task */
					char *cmd;
					char **cmdargs = NULL;
					static char bbsleepenv[20];

					/* Setup environment */
					if (twalk->envfile) {
						dbgprintf("%s -> Loading environment from %s area %s\n", 
							twalk->key, expand_env(twalk->envfile), 
							(twalk->envarea ? twalk->envarea : ""));
						loadenv(expand_env(twalk->envfile), twalk->envarea);
					}

					/* Setup BBSLEEP to match the interval */
					sprintf(bbsleepenv, "BBSLEEP=%d", twalk->interval);
					putenv(bbsleepenv);

					/* Setup command line and arguments */
					cmdargs = setup_commandargs(twalk->cmd, &cmd);

					/* Point stdout/stderr to a logfile, if requested */
					if (twalk->logfile) {
						char *logfn = expand_env(twalk->logfile);

						dbgprintf("%s -> Assigning stdout/stderr to log '%s'\n", twalk->key, logfn);

						freopen(logfn, "a", stdout);
						freopen(logfn, "a", stderr);
					}

					/* Go! */
					dbgprintf("%s -> Running '%s', BBHOME=%s\n", twalk->key, cmd, xgetenv("BBHOME"));
					execvp(cmd, cmdargs);

					/* Should never go here */
					errprintf("Could not start task %s using command '%s': %s\n", 
						   twalk->key, cmd, strerror(errno));
					exit(0);
				}
				else if (twalk->pid == -1) {
					/* Fork failed */
					errprintf("Fork failed!\n");
					twalk->pid = 0;
				}
				else {
					if (twalk->group) twalk->group->currentuse++;
					if (verbose) errprintf("Task %s started with PID %d\n", twalk->key, (int)twalk->pid);
				}
			}
			else if (twalk->pid > 0) {
				dbgprintf("Task %s active with PID %d\n", twalk->key, (int)twalk->pid);
				if (twalk->maxruntime && ((now - twalk->laststart) > twalk->maxruntime)) {
					errprintf("Killing hung task %s (PID %d) after %d seconds\n",
						  twalk->key, (int)twalk->pid,
						  (now - twalk->laststart));
					kill(twalk->pid, (twalk->beingkilled ? SIGKILL : SIGTERM));
					twalk->beingkilled = 1; /* Next time it's a real kill */
				}
			}
		}

		sleep(5);
	}

	/* Shutdown running tasks */
	for (twalk = taskhead; (twalk); twalk = twalk->next) {
		if (twalk->pid) kill(twalk->pid, SIGTERM);
	}

	if (pidfn) unlink(pidfn);

	return 0;
}

